memgpt/agent.py

```diff
@@ -103,30 +109,69 @@ def initialize_memory(ai_notes, human_notes):
     return memory
 
 
-def construct_system_with_memory(system, memory, memory_edit_timestamp, archival_memory=None, recall_memory=None, include_char_count=True):
-    full_system_message = "\n".join(
-        [
-            system,
-            "\n",
-            f"### Memory [last modified: {memory_edit_timestamp.strip()}]",
-            f"{len(recall_memory) if recall_memory else 0} previous messages between you and the user are stored in recall memory (use functions to access them)",
-            f"{len(archival_memory) if archival_memory else 0} total memories you created are stored in archival memory (use functions to access them)",
-            "\nCore memory shown below (limited in size, additional information stored in archival / recall memory):",
-            f'<persona characters="{len(memory.persona)}/{memory.persona_char_limit}">' if include_char_count else "<persona>",
-            memory.persona,
-            "</persona>",
-            f'<human characters="{len(memory.human)}/{memory.human_char_limit}">' if include_char_count else "<human>",
-            memory.human,
-            "</human>",
-        ]
-    )
+def construct_system_with_memory(

```

Comment:
this feels to me like an increase in complexity to a very core component of the overall agent.


I think the templated core memory feature is a cool idea, but I think the templating handling should be encapsulated into a class and then passed into this function.

Something like
`def construct_system_with_memory(system: SystemPromptHandler, ...`

then the full system message gets handled via

`system.get_message()` or something similar.

This way if someone wants to experiment with a different approach, they can simply swap out the SystemHandler class, rather than needing to work around the templating system here

Reply:
I think you are right about that. I wrote the code after I hadn't slept for a night. 😄